### ISSUE TITLE
Microphone support

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#12336]
A few games, including the PSP Camera game Invizimals, use a microphone.

This issue is simply here to track the work of supporting microphones in PPSSPP.

**Comment by Panderner** [comment]
iOS and Android devices has a built-in in microphone for most devices but for computers requires external microphone unless some laptops has internal microphone

**Comment by Panderner** [comment]
And add microphone permission on Android that is missing on PPSSPP

**Comment by Panderner** [comment]
VoIP apps like Skype, Discord, Viber, Facebook Messenger and etc. Some games Talking Tom & Friends franchise, Pou, Some Games has Voice Chat Features Like PUBG, Call of Duty, Mobile Legends, Lifeafter, etc. 

The Allow (Any App Name) To Record Audio? Pop ups when the microphone permission is not enabled.
Pou For example:
![Screenshot_2020-03-11-20-49-23-96](https://user-images.githubusercontent.com/37503397/76418877-5ea34480-63da-11ea-9361-7d41666b14df.png)
And Microphone Permission Side Bar like this:
![IMG_20200311_205526](https://user-images.githubusercontent.com/37503397/76419253-fd2fa580-63da-11ea-8b93-e2be86c8e943.jpg)
This feature is only present for Android Versions 6.0 Marshmallow and newer, Android Versions Older than 6.0 Marshmallow the microphone feature always enabled

**Comment by hrydgard** [comment]
Yes that's not an issue, implementing the PSP microphone APIs well enough that the games are happy with them is.

**Comment by adenovan** [comment]
what do you prefer we use on android ? Oboe , AAudio , OpenSL ES?

**Comment by Florin9doi** [comment]
The same API used for output? You may start with this: https://github.com/Florin9doi/ppsspp/commit/d6e47a2f208c5483f0b680dade5574e4451d73e3

**Comment by adenovan** [comment]
cool some part of audio recording already implemented in ppsspp so atm its only on android ? and we have many unimplemented api there , anyone can confirm different psp camera have different api implementation ?

**Comment by Florin9doi** [comment]
Is my fork and it isn't merged. I've started it a while ago, but I didn't figure it out how to complete it.

Yes. There are 3 different API:
- sceUsbCam (GoCam!: PSP-300/PSP-450x)
- sceUsbMic (Talkman: PSP-240X)
- sceAudio (headphones)

Audio Mechanica V8d Source Code: https://www.brewology.com/downloads/download.php?id=9420&mcid=1

**Comment by Gianluca2101** [comment]
I really appreciate your work , can't wait for the microphone access!

**Comment by Panderner** [comment]
I know Drastic a Nintendo DS Emulator that supports microphone feature and it's used for some games (Mike's Stage on WarioWare Touched and naming your Dog using your voice in Nintendogs)
Citra Emulator on Android has recently added microphone support on official Build.
but how about a PSP? it has microphone built-in on PSP-3000 and some games that requires Microphone (TalkMan).

**Comment by Panderner** [comment]
#13229 is done for windows but how about android?

**Comment by DinisB** [comment]
Can you add microphone to PPSSPP android? I am stuck in the StingWing part of Invizimalz because of the microphone

**Comment by hzl12** [comment]
please i just asking, bcs on my ppsspp on android and Pc it won't work. or it's only work on gold? so i just confused how audiomechanica work, really need to know. i really want to play talkman. please

**Comment by unknownbrackets** [comment]
There's no functionality difference with gold.  Microphone is not fully implemented yet.  The way to get this to work involves writing code in C++.

-[Unknown]

**Comment by hzl12** [comment]
i see. but i don't understand how to writing the codes on ppsspp. bcs i'm not a programer. maybe you have link of tutorial for me to follow. but thanks for your respons before.

**Comment by DinisB** [comment]
maybe the source code (if ppsspp has one)

**Comment by shenweip** [comment]
@hzl12 Why do you want to play talkman with audiochanica? Wonder you can play talkman with audiochanica on a real PSP? If you   want to play on PC, you just need to run it without audiochanica, I think it would work. Of course, you are supposed to have a real Microphone and update ppsspp to the latest git build.

**Comment by hzl12** [comment]
i love to learn language, and i think talkman game is a one of a good source to learn. real microphone yeah, i'll try, thank you very much.

**Comment by MojoJojoDojo** [comment]
Too bad the multiplayer games only used microphone on infrastructure and not ad-hoc.
Unless its possible to move the functionality to ad-hoc lol.

**Comment by zackzick2** [comment]
I hope microphone ppsspp will support to invizimals series. Thank you

**Comment by hrydgard** [comment]
Yeah apparently this works now for some cases but not invizimals, which is odd.

**Comment by zackzick2** [comment]
Ow please improve the function of microphone to support invizimals game series 😁👍

**Comment by MojoJojoDojo** [comment]
Technically does not work on the infrastructure test builds by AdamN but not sure if its the microphone not implemented correctly or not.

**Comment by realriccio** [comment]
It would be great for this to be implemented also on linux, since the Steam Deck has a dedicated microphone that PPSSPP is currently unable to use.

**Comment by wutend159** [comment]
Would be great if we could implement it into iOS now that PPSSPP is on the AppStore

**Comment by hrydgard** [comment]
Linked it from #19141

**Comment by anr2me** [comment]
Just a reference to a cross-platform audio library https://github.com/mackron/miniaudio

**Comment by alexyz-0** [comment]
Now that there are stick-on usb-c cameras on the market (thanks to the nintendo switch 2), a pc handheld like the steam deck is arguably the best way to play invizimals. now a linux implementation for the microphone support would be great.

I tried using the windows version of ppsspp via proton as a workaround for invizimals that require the microphone but then obviously the camera doesn't work.
